lib/component.js

```diff
@@ -805,6 +817,97 @@ class Component extends WebComponent {
     return this._attrs;
   }
 
+  _syncParams() {
+    // maintain local validated map where all schema keys are defined
+    this._paramTypes = {};
+    const paramTypes = this.constructor.paramTypes || {};
+
+    for (const param of Object.keys(paramTypes)) {
+      // convert type shorthand to object
+      let paramType = paramTypes[param];
+      if (PARAM_TYPE_DEFAULTS.has(paramType)) {
+        paramType = {type: paramType};
+      }
+
+      // Ensure param type is valid
+      const type = paramType.type;
+      if (!PARAM_TYPE_DEFAULTS.has(type)) {
+        const typeString = typeof type === `function` ? type.name : String(type);
+        throw new Error(
+          `Invalid type: ${typeString} for attr: ${param} in paramType. ` +
+            `Only (${Array.from(PARAM_TYPE_DEFAULTS.keys())
+              .map((v) => `'${v.name}'`)
+              .join(` | `)}) is valid.`,
+        );
+      }
+
+      if (paramType.default && paramType.required) {
+        throw new Error(`${this}: param '${param}' cannot have both required and default`);
+      }
+
+      if ([Object, Array, Map, Set].includes(type) && !paramType.shouldUpdate) {
+        throw new Error(
+          `Object like params '${param}' on ${this.constructor.name} must have 'shouldUpdate' function passed in`,
+        );
+      }
+
+      const paramSchemaObj = {
+        type,
+        default: paramType.default ? paramType.default : PARAM_TYPE_DEFAULTS.get(type),
+        required: paramType.required ? paramType.required : false,
+        // by default use === for comparision
+        shouldUpdate: paramType.shouldUpdate || ((oldVal, newVal) => oldVal !== newVal),
+      };
+
+      // set default value for the params
+      this.params[param] = paramSchemaObj.default;
+
+      this._paramTypes[param] = paramSchemaObj;
+    }
+
+    const self = this;
+    // vuejs like this.data.$data to reference "real" value of params
+    Object.defineProperty(this.params, `$params`, {
+      enumerable: false,
+      value: {},
+    });
+    const paramDescription = Object.keys(paramTypes)
+      .map((param) => [
+        param,
+        {
+          set(p) {
+            if (self._paramTypes[param].required && (p === undefined || p === null)) {
+              throw new Error(`${param} is required, ${p} passed in`);
+            }
+            if (self._paramTypes[param].shouldUpdate(this.$params[param], p)) {
+              this.$params[param] = p;
+              self.enqueueUpdate();
+            }
+          },
+          get() {
+            return this.$params[param];
+          },
+        },
+      ])
+      .reduce((acc, cur) => {
+        acc[cur[0]] = cur[1];
+        return acc;
+      }, {});
+    Object.defineProperties(this.params, paramDescription);
+    return this._paramTypes;
+  }
+
+  enqueueUpdate() {
+    if (this._queuedUpdate) {
+      return;
+    }
+    this._queuedUpdate = true;
+    queueMicrotask(() => {
```

Comment:
`queueMicrotask` will ensure at the end of the tick it's being executed. So we could batch all update to one single `this.update` call

Reply:
browser support for queueMicrotask could be as old as Chrome71 

https://developer.mozilla.org/en-US/docs/Web/API/queueMicrotask

could add tests in the browser test suite to verify compatibility.

and core-js provides polyfill as well https://github.com/zloirock/core-js#queuemicrotask

Reply:
It was introduced to Safari in version 12.1.

AFAICT, we don't use `core-js`. If babel can handle it, then there shouldn't be a problem.

I guess just make sure browser tests cover it.